zebrad/src/components/mempool/storage/verified_set.rs

```diff
@@ -0,0 +1,186 @@
+///
+/// - the transparent outpoints spent by transactions in the mempool
+/// - the Sprout nullifiers revealed by transactions in the mempool
+/// - the Sapling nullifiers revealed by transactions in the mempool
+/// - the Orchard nullifiers revealed by transactions in the mempool
+#[derive(Default)]
+pub struct VerifiedSet {
+    /// The set of verified transactions in the mempool.
+    transactions: VecDeque<UnminedTx>,
+
+    /// The set of spent out points by the verified transactions.
+    spent_outpoints: HashSet<transparent::OutPoint>,
+
+    /// The set of revealed Sprout nullifiers.
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+
+    /// The set of revealed Sapling nullifiers.
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+
+    /// The set of revealed Orchard nullifiers.
+    orchard_nullifiers: HashSet<orchard::Nullifier>,
+}
+
+impl VerifiedSet {
+    /// Returns an iterator over the transactions in the set.
+    pub fn transactions(&self) -> impl Iterator<Item = &UnminedTx> + '_ {
+        self.transactions.iter()
+    }
+
+    /// Returns the number of verified transactions in the set.
+    pub fn transaction_count(&self) -> usize {
+        self.transactions.len()
+    }
+
+    /// Returns `true` if the set of verified transactions contains the transaction with the
+    /// specified `id.
+    pub fn contains(&self, id: &UnminedTxId) -> bool {
+        self.transactions.iter().any(|tx| &tx.id == id)
+    }
+
+    /// Returns `true` if the given `transaction` has any spend conflicts with transactions in the
+    /// mempool.
+    ///
+    /// Two transactions have a spend conflict if they spent the same UTXO or if they reveal the
+    /// same nullifier.
+    pub fn has_spend_conflicts(&self, unmined_tx: &UnminedTx) -> bool {
+        let tx = &unmined_tx.transaction;
+
+        Self::has_conflicts(&self.spent_outpoints, tx.spent_outpoints())
+            || Self::has_conflicts(&self.sprout_nullifiers, tx.sprout_nullifiers().copied())
+            || Self::has_conflicts(&self.sapling_nullifiers, tx.sapling_nullifiers().copied())
+            || Self::has_conflicts(&self.orchard_nullifiers, tx.orchard_nullifiers().copied())
+    }
+
+    /// Clear the set of verified transactions.
+    ///
+    /// Also clears all internal caches.
+    pub fn clear(&mut self) {
+        self.transactions.clear();
+        self.spent_outpoints.clear();
+        self.sprout_nullifiers.clear();
+        self.sapling_nullifiers.clear();
+        self.orchard_nullifiers.clear();
+    }
+
+    /// Insert a transaction into the set.
+    ///
+    /// # Correctness
+    ///
+    /// The transaction is assumed to not have spend conflicts with any other transaction already
+    /// in the set. Use [`Self::has_spend_conflicts`] to ensure this before calling this method.
+    pub fn insert(&mut self, transaction: UnminedTx) {
+        self.cache_outputs_from(&transaction.transaction);
+        self.transactions.push_front(dbg!(transaction));
```

Comment:
Nitpick: Leftover `dbg!`
```suggestion
        self.transactions.push_front(transaction);
```